spring-kafka/src/main/java/org/springframework/kafka/config/ShareKafkaListenerContainerFactory.java

```diff
@@ -0,0 +1,221 @@
+ * the classic consumer model.
+ *
+ * @param <K> the key type
+ * @param <V> the value type
+ *
+ * @author Soby Chacko
+ * @since 4.0
+ */
+public class ShareKafkaListenerContainerFactory<K, V>
+		implements KafkaListenerContainerFactory<ShareKafkaMessageListenerContainer<K, V>>, ApplicationEventPublisherAware, ApplicationContextAware {
+
+	protected final LogAccessor logger = new LogAccessor(LogFactory.getLog(getClass()));
+
+	private final ContainerProperties containerProperties = new ContainerProperties((Pattern) null);
+
+	private ShareConsumerFactory<? super K, ? super V> shareConsumerFactory;
+
+	private @Nullable Boolean autoStartup;
+
+	private @Nullable Integer phase;
+
+	private @Nullable ApplicationEventPublisher applicationEventPublisher;
+
+	private @Nullable ApplicationContext applicationContext;
+
+	/**
+	 * Construct an instance with the provided consumer factory.
+	 * @param shareConsumerFactory the share consumer factory
+	 */
+	public ShareKafkaListenerContainerFactory(ShareConsumerFactory<K, V> shareConsumerFactory) {
+		this.shareConsumerFactory = shareConsumerFactory;
+	}
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) {
+		this.applicationContext = applicationContext;
+	}
+
+	/**
+	 * Get the share consumer factory.
+	 * @return the share consumer factory
+	 */
+	public ShareConsumerFactory<? super K, ? super V> getShareConsumerFactory() {
+		return this.shareConsumerFactory;
+	}
+
+	/**
+	 * Set the share consumer factory to use for creating containers.
+	 * @param shareConsumerFactory the share consumer factory
+	 */
+	public void setShareConsumerFactory(ShareConsumerFactory<? super K, ? super V> shareConsumerFactory) {
+		this.shareConsumerFactory = shareConsumerFactory;
+	}
+
+	/**
+	 * Set whether containers created by this factory should auto-start.
+	 * @param autoStartup true to auto-start
+	 */
+	public void setAutoStartup(Boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	/**
+	 * Set the phase in which containers created by this factory should start and stop.
+	 * @param phase the phase
+	 */
+	public void setPhase(Integer phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+		this.applicationEventPublisher = applicationEventPublisher;
+	}
+
+	/**
+	 * Get the container properties.
+	 * @return the container properties
+	 */
+	public ContainerProperties getContainerProperties() {
+		return this.containerProperties;
+	}
+
+	@Override
+	public ShareKafkaMessageListenerContainer<K, V> createListenerContainer(KafkaListenerEndpoint endpoint) {
+		ShareKafkaMessageListenerContainer<K, V> instance = createContainerInstance(endpoint);
+		JavaUtils.INSTANCE
+				.acceptIfNotNull(endpoint.getId(), instance::setBeanName);
+		if (endpoint instanceof AbstractKafkaListenerEndpoint) {
```

Comment:
This could be pattern expression:
```
if (endpoint instanceof AbstractKafkaListenerEndpoint abstractKafkaListenerEndpoint) {
```
And not cast in the next line.